File: 167.two-sum-ii-input-array-is-sorted.py
#
# @lc app=leetcode id=167 lang=python3
#
# [167] Two Sum II - Input Array Is Sorted
#

# @lc code=start
import re


class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        l, r = 0, len(numbers)-1
        while l <r:
            s = numbers[l] + numbers[r]
            if s== target:
                return [l+1,r+1]
            elif s<target:
                l+=1
            else:
                r-=1


# @lc code=end


# class Solution:
#     def twoSum(self, numbers: List[int], target: int) -> List[int]:
#         dic = {}
#         for i, val in enumerate(numbers):
#             if target-val in dic:
#                 return [dic[target-val]+1, i+1]
#             dic[val] = i

File: 167.two-sum-ii-input-array-is-sorted.cpp
/*
 * @lc app=leetcode id=167 lang=cpp
 *
 * [167] Two Sum II - Input Array Is Sorted
 */
#include "bits/stdc++.h"

using namespace std;

// @lc code=start
// two pointers
class Solution {
public:
  vector<int> twoSum(vector<int> &numbers, int target) {
    int n = numbers.size();
    int l = 0, r = n - 1;
    int sum = 0;
    while (l < r) {
      sum = numbers[l] + numbers[r];
      if (sum == target) {
        return {l + 1, r + 1};
      } else if (sum < target) {
        ++l;
      } else {
        --r;
      }
    }
    return {-1, -1};
  }
};
// @lc code=end

// map
class Solution {
public:
  vector<int> twoSum(vector<int> &numbers, int target) {
    unordered_map<int, int> m;
    for (int i = 0; i < numbers.size(); ++i) {
      if (m.find(numbers[i]) == m.end()) {
        // store the index of (target - num)
        m[target - numbers[i]] = i;
      } else {
        return {m[numbers[i]] + 1, i + 1};
      }
    }
    return {-1, -1};
  }
};